actioncable/app/assets/javascripts/action_cable/connection.coffee

```diff
@@ -89,19 +96,28 @@ class ActionCable.Connection
 
     open: ->
       ActionCable.log("WebSocket onopen event")
-      @disconnected = false
-      @subscriptions.reload()
 
-    close: ->
+      if @isProtocolSupported()
+        ActionCable.log("WebSocket using '#{@getProtocol()}' subprotocol.")
+        @disconnected = false
+        @subscriptions.reload()
+      else
+        ActionCable.log("WebSocket got '#{unsupported_protocol}' as the chosen protocol. Stopping monitor and disconnecting.")
+        @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+        @close()
+
+    close: (event) ->
       ActionCable.log("WebSocket onclose event")
-      @disconnect()
+      clientEvent = new ActionCable.Events.Disconnected(event.code, event.reason, true)
+      @disconnect(clientEvent)
 
     error: ->
       ActionCable.log("WebSocket onerror event")
-      @disconnect()
 
-  disconnect: ->
+  disconnect: (event) ->
     return if @disconnected
+    if event and not event.isRecoverable
+      @monitor.stop()
```

Comment:
Check out what I just added. I put the `start` and `stop` methods on the consumer.